server/src/pages/shot/editor.js

```diff
@@ -34,23 +51,278 @@ exports.Editor = class Editor extends React.Component {
+    let cropHeight = y2 - y1;
+    let croppedImage = document.createElement('canvas');
+    croppedImage.width = parseInt(cropWidth, 10);
+    croppedImage.height = parseInt(cropHeight, 10);
+    let croppedContext = croppedImage.getContext("2d");
+    croppedContext.drawImage(this.imageCanvas, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    croppedContext.drawImage(this.editor, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    croppedContext.globalCompositeOperation = 'multiply';
+    croppedContext.drawImage(this.highlighter, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    this.canvasWidth = cropWidth;
+    this.canvasHeight = cropHeight;
+    let img = new Image();
+    let imageContext = this.imageCanvas.getContext('2d');
+    img.crossOrigin = 'use-credentials';
+    let width = cropWidth;
+    let height = cropHeight;
+    img.onload = () => {
+      imageContext.drawImage(img, 0, 0, width, height);
+    }
+    this.imageContext = imageContext;
+    img.src = croppedImage.toDataURL("image/png");
+    this.setState({image: 'cropped'});
     this.edit();
+    this.onClickCancelCrop();
   }
 
+  onClickCancelCrop() {
+    this.cropToolBar = null;
+    this.cropper.getContext('2d').clearRect(0, 0, this.cropper.width, this.cropper.height);
+    boxes = [];
+    this.setState({tool: 'pen'});
+  }
+
+  mouseup(e) {
+    if (this.clickedArea.box == -1 && this.tempBox != null && boxes.length < 1) {
+      boxes.push(this.tempBox);
+    } else if (this.clickedArea.box != -1) {
+      let selectedBox = boxes[this.clickedArea.box];
+      if (selectedBox.x1 > selectedBox.x2) {
+        let previousX1 = selectedBox.x1;
+        selectedBox.x1 = selectedBox.x2;
+        selectedBox.x2 = previousX1;
+      }
+      if (selectedBox.y1 > selectedBox.y2) {
+        let previousY1 = selectedBox.y1;
+        selectedBox.y1 = selectedBox.y2;
+        selectedBox.y2 = previousY1;
+      }
+    }
+
+    this.clickedArea = {box: -1, pos: 'o'};
+    this.tempBox = null;
+    mousedown = false;
+  }
+
+  mousedown(e) {
+    mousedown = true;
+    this.clickedArea = this.findCurrentArea(e.offsetX, e.offsetY);
+    x1 = e.offsetX;
+    y1 = e.offsetY;
+    x2 = e.offsetX;
+    y2 = e.offsetY;
+  }
+
+  mouseout(e) {
+    if (this.clickedArea.box != -1) {
+      let selectedBox = boxes[this.clickedArea.box];
+      if (selectedBox.x1 > selectedBox.x2) {
+        let previousX1 = selectedBox.x1;
+        selectedBox.x1 = selectedBox.x2;
+        selectedBox.x2 > previousX1;
+      }
+      if (selectedBox.y1 > selectedBox.y2) {
+        let previousY1 = selectedBox.y1;
+        selectedBox.y1 = selectedBox.y2;
+        selectedBox.y2 > previousY1;
+    }
+  }
+    mousedown = false;
+    this.clickedArea = {box: -1, pos: 'o'};
+    this.tempBox = null;
+  }
+
+  findCurrentArea(x, y) {
```

Comment:
There's a pattern in `ui.js` and `uicontrol.js` for this. E.g.: https://github.com/mozilla-services/screenshots/blob/master/addon/webextension/selector/uicontrol.js#L104-L114

and: https://github.com/mozilla-services/screenshots/blob/master/addon/webextension/selector/uicontrol.js#L800-L823

This method is a bit like `ui.Box.draggerDirection` but it's got a bit too much going on. Also `draggerDirection` doesn't do math on the pixel location, but uses the fact that the box has target elements: https://github.com/mozilla-services/screenshots/blob/master/addon/webextension/selector/ui.js#L783-L799